pex/interpreter.py

```diff
@@ -183,10 +184,23 @@ def get(cls, binary=None):
         # vendored attrs distribution so that its `cache_hash=True` feature can work (see the
         # bottom of pex/third_party/__init__.py where the vendor importer is installed). We ignore
         # such adjoined `sys.path` entries to discover the true base interpreter `sys.path`.
-        pythonpath = frozenset(
-            os.environ.get("PYTHONPATH", "").split(os.pathsep) + list(third_party.exposed())
+        pythonpath = os.environ.get("PYTHONPATH")
+        internal_entries = frozenset(
+            (pythonpath.split(os.pathsep) if pythonpath else []) + list(third_party.exposed())
         )
-        sys_path = OrderedSet(item for item in sys.path if item and item not in pythonpath)
+
+        def is_internal_entry(entry):
+            # type: (str) -> bool
+            if entry in internal_entries:
+                return True
+            if not os.path.isabs(entry):
+                return False
+            for internal_entry in internal_entries:
+                if internal_entry == commonpath((internal_entry, entry)):
```

Comment:
The lack of this check was the bug. Something like `~/.pex/isolated/975c556eea71292a09d930db2ca41875066d8be6/pex/vendor/_vendored/attrs` would be (and still is) on the `sys.path` during PEX boot, but only the `~/.pex/isolated/975c556eea71292a09d930db2ca41875066d8be6` entry was in PYTHONPATH since `attrs` was imported via the `pex.third_party` custom importer, which imports from `~/.pex/isolated/975c556eea71292a09d930db2ca41875066d8be6` using the `pex.vendor._vendored.attrs` package prefix.

Reply:
The fix is good, but the comment above has some BS in it I just detected. I'll amend the comment to spell out the situation a bit better in a few hours.

Reply:
Ok, there is a deeper fix that can be made such that the old check would work. The real bug is that `third_party.exposed()` is empty when it should include attrs. This is a very old bug in the `pex.third_party` custom Importer code. I'll proceed with releasing the current fix and circle back to make the fix cleaner (it'll probably be more ugly / more code, but it will be in the right place).

Reply:
Ok, actually a clean fix. This just required removing an always un-needed check present in the original introduction of the vendoring / pex.third_party mechanism in #624 way back in 2018. See #2330.